### PR TITLE
[menu] Overhaul the component API

### LINKED ISSUE #9: [Menu] Menu click away is broken if `transitionDuration` CSS property is used
### Steps to reproduce

Link to live example: https://codesandbox.io/p/sandbox/compassionate-burnell-mns7ll

https://github.com/mui/material-ui/assets/4148577/f077aa91-e336-4c5b-bfaa-2723bd85f78d

Steps:
1. Click on the first menu button, then click anywhere else on the page. The menu will collapse.
2. Click on the second menu button, then click anywhere else on the page. The menu will **not** collapse.

### Current behavior

If I use the `transitionDuration` CSS property when styling the menu, it will not collapse when I click away from it.

### Expected behavior

The menu should always collapse, no matter how I style it.

### Context

I want the menu to collapse when I click away from it, no matter how I style it.

### Your environment

<details>
  <summary><code>npx @mui/envinfo</code></summary>

I used Chrome when testing this issue.

```
  System:
    OS: macOS 14.2.1
  Binaries:
    Node: 18.19.0 - /opt/homebrew/bin/node
    npm: 10.4.0 - /opt/homebrew/bin/npm
    pnpm: 8.6.1 - /opt/homebrew/bin/pnpm
  Browsers:
    Chrome: 121.0.6167.160
    Edge: Not Found
    Safari: 17.2.1
  npmPackages:
    @emotion/react: 11.11.3 => 11.11.3 
    @emotion/styled: 11.11.0 => 11.11.0 
    @mui/base: 5.0.0-beta.36 => 5.0.0-beta.36 
    @mui/utils: 5.15.9 => 5.15.9 
    @types/react: 18.2.55 => 18.2.55 
    react: 18.2.0 => 18.2.0 
    react-dom: 18.2.0 => 18.2.0 
    typescript: 5.3.3 => 5.3.3
```
</details>

**Search keywords**: menu click away broken css

### LINKED ISSUE #46: [Menu] Configure MenuItems not to close the Menu when clicked
https://github.com/mui/material-ui/issues/38324#issuecomment-1682293358

Consider adding a prop (such as `disableCloseOnClick`) that modifies the default MenuItem behavior.

### LINKED ISSUE #63: [base-ui][Menu] Support opening the menu and selecting an item with one click
Similarly to the native macOS menus, we could support executing opening the menu and performing actions with only one click. This requires the menu to open on `pointerdown` and menu items to execute on `pointerup`.
This way, users will be able to press the mouse button on the menu button, move to an item while still holding the button, and finally release the button over the item to execute it.

**Search keywords**:

### LINKED ISSUE #214: [Menu] Implement new API
API Doc: https://www.notion.so/mui-org/base-ui-Menu-API-b6d240669fb242b1bf86acda25e88c21

**Related issues**
Each of these related issues should be considered when designing and implementing the new Menu. When this issue is resolved, these related issues should be closed.

- #9 
- #63
- #53
- #52
- #46
- #411

### LINKED ISSUE #411: [Menu] Closing the menu with Tab doesn't restore focus to the trigger element
### Steps to reproduce

Link to live example: https://codesandbox.io/p/sandbox/unruffled-mirzakhani-8zvqrp?file=%2Fsrc%2FDemo.tsx

[menu.webm](https://github.com/mui/material-ui/assets/42154031/60f66d0e-65c6-4acf-9599-db9eb31c45ee)

### Current behavior

Pressing <kbd>Tab</kbd> closes the menu and, since it's portaled and there's no other element after it, the focus goes to `<body>`.

### Expected behavior

The focus should return to the element that triggered the menu, like how <kbd>Esc</kbd> works.

### Context

_No response_

### Your environment

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Search keywords**: base menu tab

**Search keywords**:

### PR DESCRIPTION
**Pull request by michaldudak**
Closes #214
Closes #9
Closes #46
Closes #63
Closes #411

## Preview
https://deploy-preview-468--base-ui.netlify.app/base-ui/react-menu/

## To do
- [x] #411 
- [x] #46 
- [x] #63 
- [x] #9 
- [X] Nested menus
- [x] Animations
- [x] Docs
- [x] Configure <kbd>Esc</kbd> key behavior (close all/close one submenu level)
- [x] `keepMounted` prop
- [x] Configurable delay and open on hover

### In separate PRs
- [x] #52 
- [x] #53 
- [x] #70

### CODE REVIEW

**Comment by colmtuite** [comment]
@michaldudak 

- Submenu shouldn't close when the pointer leaves its popup. It should close when focus moves to a menuitem that is not the submenu's trigger, and that is not disabled.
- When you open a submenu via pointer, focus should remain on its trigger. Pressing right should move focus into the submenu.
- 750ms typeahead threshold is too long.
- When opening via space/enter, it should trigger by keydown rather than keypress.
- Vid attached of placing pointer over a menuitem to focus it, then pressing up, and two menuitems are highlighted. Only one menuitem should ever be focused.
- Vid attached of some crazy ass bug

https://github.com/mui/base-ui/assets/805073/044b66aa-fa08-429c-87d8-51d18ee04f56
https://github.com/mui/base-ui/assets/805073/294742ba-0bcf-4034-abd8-5b5f318eac67

**Comment by michaldudak** [comment]
@colmtuite 

Fixed:
- submenus don't close on mouseout anymore
- focus stays on the highlighted submenu trigger after it opens (EDIT: there's still something weird going on on first interaction; I'm investigating it)
- changed typeahead threshold to 350ms
- bugs you recorded are fixed

**Comment by colmtuite** [comment]
@michaldudak When you close the menu via ESC, focus is not returned to the trigger.

File: packages/mui-base/src/Menu/Root/useMenuRoot.ts [hunk 1]
```diff
@@ -192,8 +195,9 @@ export function useMenuRoot(parameters: UseMenuRootParameters) {
   const typeahead = useTypeahead(floatingRootContext, {
     listRef: itemLabels,
     activeIndex,
+    resetMs: 350,
```
Comment (atomiks) [inline packages/mui-base/src/Menu/Root/useMenuRoot.ts:198]: @colmtuite  Why so short? For slow typers this seems way too fast to me. macOS' typeahead for Menus is over 500ms (not sure on the exact value, but 750ms seems like a good default).

File: packages/mui-base/src/Menu/Root/menuReducer.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    case MenuActionTypes.open: {
+      const updateHighlight = action.event instanceof KeyboardEvent;
+
+      return {
+        ...state,
+        open: true,
+        highlightedValue: updateHighlight
+          ? moveHighlight(null, 'start', state.items, state.settings)
+          : state.highlightedValue,
+      };
```
Comment (atomiks) [inline packages/mui-base/src/Menu/Root/menuReducer.ts:24]: What is the purpose of this reducer? `useListNavigation` handles the highlight (focus) all by itself, so this is likely conflicting and causing the possible bugs

Comment (michaldudak) [reply packages/mui-base/src/Menu/Root/menuReducer.ts:24]: Without handling this here I had to press the right arrow twice to actually focus a first element in the menu.

File: packages/mui-base/src/Menu/Root/useMenuRoot.ts (hunk 1 above)
Comment (colmtuite) [reply packages/mui-base/src/Menu/Root/useMenuRoot.ts:189]: iirc, @michaldudak suggested 350ms? Not sure. I don't think I had a specific preference, just that the original value felt clearly too long.

**Comment by colmtuite** [comment]
@michaldudak Clicking the trigger repeatedly should toggle the menu open/closed

**Comment by DiegoAndai** [comment]
Is the space/enter key opening broken? I remember it worked last week (coming here from the monthly meeting link 😅)


https://github.com/mui/base-ui/assets/16889233/06682b75-3472-4b5b-a71d-3935129bfc9e

**Comment by michaldudak** [comment]
It is for now, but bear in mind it's still a draft and I'm working on it :)

**Comment by colmtuite** [comment]
When you move the cursor very quickly across menu items, it seems like focus is not able to keep up.
https://github.com/mui/base-ui/assets/805073/7f1f33cb-172f-47aa-9519-bc8f76493625

**Comment by michaldudak** [comment]
> Clicking the trigger repeatedly should toggle the menu open/closed

In MacOS, the native menus open on the trigger's mousedown but close on mouseup. WDYT about adopting this behavior?

**Comment by colmtuite** [comment]
MacOS has some sort of interval. If you click, hold for 50ms, then release over the trigger, the menu remains open. If you click, hold for 500ms, then release over the trigger, the menu closes.

After chatting, we decided this might have two issues:

1. It may be problematic for some older or disabled users who can't confidently click the button fast enough.
2. There will be some perf hit to set a timer on each click.

So we decided to close the menu when you click, hold, drag, and release outside of the trigger.

**Comment by atomiks** [comment]
Some bugs:

1. The focus doesn't synchronize once pointer-events stop being blocked and the cursor is at rest

https://github.com/user-attachments/assets/68f68436-15bb-48d9-a52a-4ba61896eb37

2. The submenu is left open when quickly navigating with ArrowLeft/ArrowRight even though focus is on the parent trigger

https://github.com/user-attachments/assets/9fe69fcc-f3e4-4353-9e05-aa11860c56f3

**Comment by michaldudak** [comment]
> The focus doesn't synchronize once pointer-events stop being blocked and the cursor is at rest

This looks like a bug in Floating UI. The same behavior can be seen in https://codesandbox.io/p/sandbox/admiring-lamport-5wt3yg.

I can take a look at it, unless you have an idea how to solve it easily.

> The submenu is left open when quickly navigating with ArrowLeft/ArrowRight even though focus is on the parent trigger

This one's interesting. I was only able to reproduce it when animations are enabled, and it seems like the first element of the submenu isn't focused when the menu is open. I'm still investigating the root cause.

**Comment by atomiks** [comment]
> This looks like a bug in Floating UI. The same behavior can be seen in https://codesandbox.io/p/sandbox/admiring-lamport-5wt3yg.

I don't see the bug in that sandbox? When it rests the focus synchronizes. (I think it may be manually handled externally in the components, not the internal hooks though)

Actually I see it doesn't synchronize unless the element underneath the cursor is a submenu trigger. I think that was intentional since nothing opens on `hover` and doesn't need to/shouldn't synchronize, but it's been a while since I built that.

**Comment by michaldudak** [comment]
What do you think about not disabling pointer events in these cases? This will also help in cases where someone wants to select a menu item below a submenu trigger right after the submenu opens.

**Comment by atomiks** [comment]
- [Do not block pointer-events in safe polygon](https://github.com/mui/base-ui/pull/468/commits/15ffd83a65c771becd3e9d8ff8b0502d1da04d1c) - why was this added?

> This will also help in cases where someone wants to select a menu item below a submenu trigger right after the submenu opens.

The `requireIntent` feature of `safePolygon` handles this scenario. It prevents them from being too aggressively blocked by using a timeout

- If another button is focused instead of the parent menu trigger, and a submenu item is clicked, it returns focus to that instead of the parent. The top-level menu items aren't affected. 
- `escapeClosesParents` as a prop name sounds slightly awkward

**Comment by michaldudak** [comment]
> [Do not block pointer-events in safe polygon](https://github.com/mui/base-ui/pull/468/commits/15ffd83a65c771becd3e9d8ff8b0502d1da04d1c) - why was this added?

This is a workaround for "The focus doesn't synchronize once pointer-events stop being blocked and the cursor is at rest".

> If another button is focused instead of the parent menu trigger, and a submenu item is clicked, it returns focus to that instead of the parent. The top-level menu items aren't affected.

Could you record a video showing this issue?

**Comment by atomiks** [comment]
> Could you record a video showing this issue?

Bug seems to be gone in the latest deploy

> This is a workaround for "The focus doesn't synchronize once pointer-events stop being blocked and the cursor is at rest".

It only needs to synchronize focus for a submenu trigger, not a menu item. This is because the submenu item opens a menu on hover, so it needs to focus style applied. The Floating UI menu demo handles only that

**Comment by michaldudak** [comment]
@atomiks I restored `blockPointerEvents` and managed to solve the issue with CSS.

**Comment by atomiks** [comment]
Bug when traversing back to the submenu trigger, menu doesn't open

https://github.com/user-attachments/assets/162b1b95-74bb-4a82-9fc1-927a6b835d00

**Comment by colmtuite** [comment]
@michaldudak Feedback based on our chat:

1. Add the Arrow component
2. Consider making focus looping configurable. I vote it should be.
3. Rename `positionStrategy` to `positionMethod` across all components.

**Comment by michaldudak** [comment]
@colmtuite I did 1 and 2. `positionStrategy` rename should be done on the other components first in a separate PR.

**Comment by colmtuite** [comment]
@michaldudak In this vid, the menu is initially opening above the trigger, then quickly repositioning itself below.
https://github.com/user-attachments/assets/dee98c5c-a19b-47ab-8f8b-2c2f07b0eb30

**Comment by michaldudak** [comment]
> Bug when traversing back to the submenu trigger, menu doesn't open

@atomiks, Should be working OK now.

> In this vid, the menu is initially opening above the trigger, then quickly repositioning itself below.

@colmtuite, could you prepare a reproduction on codesandbox?

File: packages/mui-base/src/Menu/Positioner/MenuPositionerContext.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+   */
+  alignment: 'start' | 'end' | 'center';
+  arrowRef: React.MutableRefObject<Element | null>;
+  arrowUncentered: boolean;
+  arrowStyles: React.CSSProperties;
+}
+
+export const MenuPositionerContext = React.createContext<MenuPositionerContext | null>(null);
+
+MenuPositionerContext.displayName = 'MenuPositionerContext';
```
Comment (oliviertassinari) [inline packages/mui-base/src/Menu/Positioner/MenuPositionerContext.ts:21]:
```suggestion
if (process.env.NODE_ENV !== 'production') {
  MenuPositionerContext.displayName = 'MenuPositionerContext';
}
```

**Review by colmtuite — APPROVED** [review]

**Review by atomiks — APPROVED** [review]
Looks great!

**Pull request merged**